extensions/amp-story/1.0/amp-story.js

```diff
@@ -2375,6 +2367,21 @@ export class AmpStory extends AMP.BaseElement {
       nextPageEl.setAttribute(Attributes.RETURN_TO, pageToBeInsertedId);
     }
 
+    // Adjust the ad page's position in this.adPages_ array to reflect the actual.
+    const adPageIndex = this.getPageIndexById(pageToBeInsertedId);
+    if (adPageIndex != -1) {
+      this.pages_.splice(adPageIndex, 1);
+    }
+    this.pages_.splice(
+      this.getPageIndexById(pageBeforeId) + 1,
+      0,
+      pageToBeInserted
+    );
+
+    this.storeService_.dispatch(Action.INSERT_AD_PAGE, {
+      pageBeforeId,
+      pageToBeInsertedId,
+    });
```

Comment:
All the logic inside of the `INSERT_AD_PAGE` is not needed, you already have the pages array with all the IDs.
Plz delete the action, and all you need is:

```
    this.storeService_.dispatch(Action.SET_PAGE_IDS, this.pages_.map((el) => el.id));
```